File: mobile/ios/FindlyKit/Sources/FindlyKit/DesignSystem/Components/FindlyDropdownField.swift
import SwiftUI

/// specs/010-app-shell-and-screen-ux.md §1.2/§4.2, specs/004-ios-client.md §2.3 (010-shell
/// addition) — one selectable value in a `FindlyDropdownField`. Generic over `Value` so the same
/// component serves any single-select field (today: device sync interval, 001 §1.4's seven-value
/// set); it carries no knowledge of what those values mean. `disabledReason` is shown alongside a
/// disabled option instead of silently omitting it — §4.2: "render disabled with the limit as the
/// reason", so the option stays visible (never hidden) but not selectable.
public struct FindlyDropdownOption<Value: Hashable>: Identifiable, Equatable {
    public let value: Value
    public let title: String
    public let isEnabled: Bool
    public let disabledReason: String?

    public var id: Value { value }

    public init(value: Value, title: String, isEnabled: Bool = true, disabledReason: String? = nil) {
        self.value = value
        self.title = title
        self.isEnabled = isEnabled
        self.disabledReason = disabledReason
    }
}

/// specs/010-app-shell-and-screen-ux.md §1.2/§4.2, specs/004-ios-client.md §2.3 — a labeled
/// single-select field presented via the platform-native menu (010 §4.2: "exposed dropdown menu
/// on Android, `Menu`/`Picker` presentation on iOS"). Stateless/presentational like every other
/// design-system component: it renders whatever `options` the caller supplies and reports the
/// selection via `onSelect` — zero knowledge of view models, networking, `features`, or which
/// value is "the sync interval". The caller (a pure plan type, e.g. `SyncIntervalDropdownPlan`)
/// owns which options are enabled/disabled and why.
///
/// Reads `@Environment(\.theme)` only — no literal `Color(...)`, `.font(.system(size:))`, or
/// hardcoded point size (004 §2.3's contract, which this 010-shell addition joins).
public struct FindlyDropdownField<Value: Hashable>: View {
    @Environment(\.theme) private var theme
    @Environment(\.isEnabled) private var isEnabled
    private let label: String
    private let options: [FindlyDropdownOption<Value>]
    private let selection: Value
    private let onSelect: (Value) -> Void

    public init(
        label: String,
        options: [FindlyDropdownOption<Value>],
        selection: Value,
        onSelect: @escaping (Value) -> Void
    ) {
        self.label = label
        self.options = options
        self.selection = selection
        self.onSelect = onSelect
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: theme.spacing.xs) {
            Text(label)
                .font(theme.typography.labelSmall.font)
                .tracking(theme.typography.labelSmall.tracking)
                .foregroundColor(theme.onSurfaceMuted)
            Menu {
                ForEach(options) { option in
                    Button {
                        onSelect(option.value)
                    } label: {
                        Text(menuItemTitle(for: option))
                    }
                    .disabled(!option.isEnabled)
                }
            } label: {
                fieldBox
            }
            .disabled(!isEnabled)
            .accessibilityLabel(label)
            .accessibilityValue(currentTitle)
        }
    }

    private func menuItemTitle(for option: FindlyDropdownOption<Value>) -> String {
        guard !option.isEnabled, let reason = option.disabledReason else { return option.title }
        return "\(option.title) — \(reason)"
    }

    private var fieldBox: some View {
        HStack(spacing: theme.spacing.sm) {
            Text(currentTitle)
                .font(theme.typography.bodyLarge.font)
                .foregroundColor(isEnabled ? theme.colors.onSurface : .findlyTextFieldDisabledText)
            Spacer(minLength: theme.spacing.sm)
            // A plain glyph, not an SF Symbol — same convention `StatusChip` already uses for its
            // status glyphs, so no new asset/import is needed for one static chevron.
            Text("▾")
                .font(theme.typography.bodyLarge.font)
                .foregroundColor(theme.outlineStrong)
        }
        .padding(.horizontal, theme.spacing.sm)
        .frame(height: 52)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(isEnabled ? theme.colors.surfaceVariant : .findlyTextFieldDisabledFill)
        .clipShape(RoundedRectangle(cornerRadius: theme.corner.md))
        .overlay(
            RoundedRectangle(cornerRadius: theme.corner.md)
                .strokeBorder(isEnabled ? theme.outlineStrong : .findlyTextFieldDisabledBorder, lineWidth: 1.5)
        )
    }

    private var currentTitle: String {
        options.first(where: { $0.value == selection })?.title ?? ""
    }
}

#Preview("FindlyDropdownField — light") {
    VStack(spacing: 16) {
        FindlyDropdownField(
            label: "Sync interval",
            options: SyncIntervalDropdownPlan.options(minSyncIntervalMinutes: 5),
            selection: 15,
            onSelect: { _ in }
        )
        FindlyDropdownField(
            label: "Sync interval",
            options: SyncIntervalDropdownPlan.options(minSyncIntervalMinutes: 15),
            selection: 30,
            onSelect: { _ in }
        )
    }
    .padding()
    .background(Theme.light.colors.surface)
    .environment(\.theme, .light)
}

#Preview("FindlyDropdownField — dark") {
    VStack(spacing: 16) {
        FindlyDropdownField(
            label: "Sync interval",
            options: SyncIntervalDropdownPlan.options(minSyncIntervalMinutes: 5),
            selection: 15,
            onSelect: { _ in }
        )
    }
    .padding()
    .background(Theme.dark.colors.surface)
    .environment(\.theme, .dark)
    .preferredColorScheme(.dark)
}

File: mobile/android/app/src/main/java/com/findly/android/ui/designsystem/components/FindlyDropdownField.kt
package com.findly.android.ui.designsystem.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.disabled
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import com.findly.android.ui.designsystem.FindlyTheme
import com.findly.android.ui.designsystem.token.TextFieldDisabledBorder
import com.findly.android.ui.designsystem.token.TextFieldDisabledFill
import com.findly.android.ui.designsystem.token.TextFieldDisabledText

/**
 * One presentable option in a [FindlyDropdownField]. [enabled]/[disabledReason] let a caller
 * pre-disable a value the current plan can't use — specs/010-app-shell-and-screen-ux.md §4.2:
 * "Values below `features.limits.minSyncIntervalMinutes` render disabled with the limit as the
 * reason" — while still *presenting* every allowed value (010 §9: the server remains the sole
 * source of truth for validation; this is presentation only, never the only guard).
 */
data class FindlyDropdownOption<T>(
    val value: T,
    val label: String,
    val enabled: Boolean = true,
    val disabledReason: String? = null,
)

/**
 * A labeled single-select exposed dropdown (specs/010-app-shell-and-screen-ux.md §1.2/§4.2, added
 * to the 003 §4.3 component contract in the same PR as that spec) — the device sync-interval
 * control over the 001 §1.4 value set. Stateless/presentational like every other design-system
 * component (003 §4.3): the caller supplies [selected]/[options] and reacts to a choice in
 * [onSelect] — this component never calls a network API or reads a ViewModel, and selecting an
 * option **commits immediately** (010 §4.2: "no Save button") — that is the caller's contract to
 * honor in [onSelect], not something this component enforces itself.
 *
 * Platform-native mechanics underneath — Android's exposed-dropdown-menu idiom
 * ([DropdownMenu]/[DropdownMenuItem], 010 §4.2's "exposed dropdown menu on Android") — are this
 * component's own implementation detail, the same category [FindlyNavDrawer]'s doc describes for
 * `ModalNavigationDrawer`-class behavior: not one of the 003 §4.3 "components only" screen-level
 * exceptions, because this *is* the design system's own implementation of the control, not a
 * screen reaching past it.
 *
 * Geometry/colors deliberately mirror [FindlyTextField] (52dp field, `md` corner,
 * `surfaceVariant` fill, `outlineStrong` border, `primary` when open) so a screen combining both —
 * the §4.2 device card does, sync-interval dropdown next to the rename text field — reads as one
 * visual language. Reads only [FindlyTheme] tokens: every piece of text uses a
 * [FindlyTheme.typography] role (never a literal `TextStyle`/font size), every color is a
 * [FindlyTheme.colors] field or one of the existing disabled-state token constants
 * ([TextFieldDisabledFill]/[TextFieldDisabledBorder]/[TextFieldDisabledText], already established
 * by [FindlyTextField] — reused here rather than adding new token names, per the 003 §4.1 rule
 * that the token vocabulary is not renamed/extended without a spec PR).
 */
@Composable
fun <T> FindlyDropdownField(
    label: String,
    selected: T,
    options: List<FindlyDropdownOption<T>>,
    onSelect: (T) -> Unit,
    modifier: Modifier = Modifier,
    enabled: Boolean = true,
) {
    var expanded by remember { mutableStateOf(false) }
    val colors = FindlyTheme.colors
    val shape = RoundedCornerShape(FindlyTheme.corner.md)
    val selectedLabel = options.firstOrNull { it.value == selected }?.label.orEmpty()

    val fill: Color
    val borderColor: Color
    val textColor: Color
    when {
        !enabled -> {
            fill = TextFieldDisabledFill
            borderColor = TextFieldDisabledBorder
            textColor = TextFieldDisabledText
        }
        expanded -> {
            fill = colors.surfaceVariant
            borderColor = colors.primary
            textColor = colors.onSurface
        }
        else -> {
            fill = colors.surfaceVariant
            borderColor = colors.outlineStrong
            textColor = colors.onSurface
        }
    }

    Column(modifier = modifier.fillMaxWidth()) {
        Text(text = label.uppercase(), color = colors.subtleText, style = FindlyTheme.typography.labelSmall)

        Box {
            var fieldModifier = Modifier
                .fillMaxWidth()
                .defaultMinSize(minHeight = 52.dp)
                .clip(shape)
                .background(fill)
                .border(width = 1.5.dp, color = borderColor, shape = shape)
            fieldModifier = if (enabled) {
                fieldModifier.clickable(onClickLabel = label) { expanded = true }
            } else {
                fieldModifier.semantics { disabled() }
            }

            Row(
                modifier = fieldModifier
                    .padding(horizontal = FindlyTheme.spacing.sm, vertical = FindlyTheme.spacing.sm)
                    .semantics { contentDescription = label },
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.SpaceBetween,
            ) {
                Text(text = selectedLabel, color = textColor, style = FindlyTheme.typography.bodyLarge)
                Text(text = if (expanded) "▲" else "▼", color = textColor, style = FindlyTheme.typography.labelSmall)
            }

            DropdownMenu(
                expanded = expanded,
                onDismissRequest = { expanded = false },
                containerColor = colors.surface,
            ) {
                options.forEach { option ->
                    DropdownMenuItem(
                        text = {
                            Column {
                                Text(
                                    text = option.label,
                                    color = if (option.enabled) colors.onSurface else colors.subtleText,
                                    style = FindlyTheme.typography.bodyLarge,
                                )
                                if (!option.enabled && option.disabledReason != null) {
                                    Text(
                                        text = option.disabledReason,
                                        color = colors.subtleText,
                                        style = FindlyTheme.typography.labelSmall,
                                    )
                                }
                            }
                        },
                        enabled = option.enabled,
                        onClick = {
                            expanded = false
                            onSelect(option.value)
                        },
                    )
                }
            }
        }
    }
}
